ssl/s3_lib.c

```diff
@@ -4300,7 +4299,7 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
             if (prefer_sha256) {
                 const SSL_CIPHER *tmp = sk_SSL_CIPHER_value(allow, ii);
 
-                if (ssl_md(tmp->algorithm2) == mdsha256) {
+                if (EVP_MD_is_a(ssl_md(s->ctx, tmp->algorithm2), "SHA-256")) {
```

Comment:
I think there is a possible corner case problem here. Imagine we are using the default libctx but the default provider is not loaded into it and there is no implementation of SHA-256. However we also have an engine that *does* provide it. In that case ssl_evp_md_fetch() will have decided to use the legacy EVP_MD object for sha256, and `EVP_MD_is_a()` is going to fail to match it because the "SHA-256" name (or OSSL_DIGEST_NAME_SHA2_256) does not appear in the namemap.

Reply:
Maybe this could also happen even if the default provider is available in the default context but we have not yet fetched any digests from it and all the digests we need are engine provided. In that case the namemap is also not populated (I think??)

Reply:
In that case, I would use `OBJ_nid2sn(NID_sha256)` to be sure to get the legacy name.

Reply:
Add a TODO comment that we need to change that when legacy stuff is going away.

Reply:
Done

Reply:
Er...  hmm.  So maybe `ssl_md` should do an implicit fetch of the algorithm it gets?

Reply:
Oh, I got in a little late...  sorry 'bout that